### PR TITLE
kmsdrm regression fixes

### PR DESCRIPTION
**Pull request by inolen**
Hi,

I last updated SDL at commit 850d9c8c0daff485a07f065846ef0b00d067b258 at which point the KMSDRM backend was working well for my use case (https://redream.io). However, after updating to the latest I was having a number of issues:

* The mode requested in KMSDRM_SetDisplayMode wasn't updating the new surface_w / surface_h variables which caused the surface being created to be out of sync with the mode set. This fix is fairly straight forward.
* The mode requested in KMSDRM_SetDisplayMode wasn't being honored when calling SDL_SetWindowFullscreen(..., SDL_WINDOW_FULLSCREEN). KMSDRM_ReconfigureWindow was using the previous window size to find a mode and overriding the requested mode with that. My patch contains a fix for my particular case, but as mentioned in the commit message I'm still worried by the behavior of overriding the user's requested mode in other cases. AFAIK you should be able to toggle fullscreen with SDL_SetWindowFullscreen without having to call into SDL_SetWindowDisplayMode on other backends.
* There was previously code in KMSDRM_SetDisplayMode to defer the surface creation until the surface was first presented so that calls to SDL_GL_GetCurrentContext and SDL_EGL_MakeCurrent were accessing the correct thread-local state in the case of multiple threads / contexts. This was removed which was causing multi-threaded rendering to implode. The commit that removed it was somewhat vague, so I'm not sure of the correct fix here. My commit reverts the removal of this code, but there may need to be some more work to split up the creation of the surfaces / binding of the context in case the surfaces truly do need to be created immediately inside of KMSDRM_SetDisplayMode.

There are more notes in each commit message tracking the changes that led to each regression. While these commits fix my particular use case, there's likely some discussion to be had regarding how to correctly address all use cases.

### CODE REVIEW

**Comment by icculus** [comment]
Gonna sanity check this shortly, but at first glance it looks reasonable.

**Comment by inolen** [comment]
Force pushed just now to rebase and fix a compile error on the first commit which was resolved on one of the subsequent commits.

**Comment by sezero** [comment]
gcc emits a (minor) warning:
```
src/video/kmsdrm/SDL_kmsdrmvideo.c: In function 'KMSDRM_ReconfigureWindow':
src/video/kmsdrm/SDL_kmsdrmvideo.c:1310: warning: unused variable 'windata'
```

**Comment by inolen** [comment]
Force pushed fix for warning.

**Comment by vanfanel** [comment]
Hi. These changes break scaling in a lot of programs, like:
https://github.com/8bitbubsy/pt2-clone
https://github.com/NagyD/SDLPoP

**PLEASE NOTE that KMSDRM makes non-windowed windows look fullscreen. That feature has to stay: windows filling only a small part of the screen make no sense on KMSDRM. This was discussed with @slouken time ago. **

Before doing commits on such a delicate backend, you should not only test your user case, which is a private and closed-source emulator anyway, but do extensive testing as I do before considering that things are OK.

As for what have you broken, you have to find out yourself because I don't have the time to dive into your changes. But new release should not contain these changes because they break the KMSDRM backend.

**Comment by inolen** [comment]
Hi,

If you can point to what's causing issues I'd be glad to discuss my reasoning and we can find a solution that works for all use cases.

**Comment by vanfanel** [comment]
@inolen Revert all those changes. ALL of them. Then you can start doing one commit at a time, and I will test each individual commit.
As things are, you have broken the backend and you have to fix it, not me.
When I last touched it, it worked for all software I threw at it. And it has been working fine for months until you did those commits. So it's your responsability, not mine.
I count on the main devs to not do a broken release.

In other words: I am not putting hours, maybe days or weeks, to find out why your code breaks something previously working.

**Comment by slime73** [comment]
Can you at least describe how to reproduce the breaking issues you're seeing? If no one else aside from you knows what to test, then the chance of fixing the issues you're running into aren't very high.

**Comment by vanfanel** [comment]
@slime73 Simply run the games in fullscreen and windowed mode using the KMSDRM backend. There's nothing special to do, the scaling in all 2D games is broken.

**Comment by inolen** [comment]
Without being told what code is causing issues yet, if I had to guess it would be:

https://github.com/libsdl-org/SDL/pull/4396/commits/270cd3a79e49a18f3658eb8b7a1877a3840effad

The previous code was ignoring the mode set by SDL_SetWindowDisplayMode when SDL_SetWindowFullscreen(..., SDL_WINDOW_FULLSCREEN) was called, and instead tried to find a mode close to the previous window size. From what I'm picking up here, this may have been added to help make massage windowed apps into running correctly on this backend.

If that's the case, this kind of logic (finding an appropriate default mode based on the window size) may be appropriate to move into CreateWindow / SetWindowSize, but as it was it was it was breaking the normal path where nothing was being massaged, and SDL_SetWindowDisplayMode  -> SDL_SetWindowFullscreen was used to perform an explicit mode set.

**Comment by vanfanel** [comment]
@inolen Don't wait for me to tell you what you broke. Just take your commits back and send at a time.
You have your list of software you have broken by misunderstanding how the backend works.
I will repeat. You have broken at least these programs, that worke fine previously:
https://github.com/8bitbubsy/pt2-clone
https://github.com/NagyD/SDLPoP

You have also broken this:
https://github.com/sm64pc/sm64ex

Testing is as easy as building those and running them normally, both in windowed and fullscreen modes.

**Comment by inolen** [comment]
@slouken carrying over from https://github.com/libsdl-org/SDL/commit/fc508eabe2ed808c4b2258bdbbab5940b8c9a0f8#commitcomment-51848146. Yes, I've got a tentative fix together but need to test some other applications tomorrow.

**Comment by inolen** [comment]
As an update, I was working on some fixes to issues I noticed while trying to get redream running in Windowed mode. However, after getting done with them I wasn't sure how they'd relate to other windowed apps (the main issue I had was that SetWindowSize was still broken with multithreaded rendering, I had only updated SetDisplayMode to defer surface recreation).

I tested SDLPoP and sm64ex, both of which seemed to work the same before and after my changes on my RPI4, windowed and fullscreen.

IMG_20210608_120654

IMG_20210608_120723

On Prince of Persia I do get different aspect ratios when running in "windowed" mode rather than fullscreen, but I get the same behavior before these changes as well.

**Comment by icculus** [comment]
> All in all, NOTHING seems to work well with these commits. I hope your closed-source commercial emulator is doing well.

@vanfanel I understand that bugs are frustrating, and we can work through fixing any issues, but please don't talk to people like this.

**Comment by vanfanel** [comment]
@icculus I am sorry about that tone, but it's not the first time the same person starts making commits without testing software beyond his own closed-source usercase.
I have put many hours of work in the KMSDRM backend, and after I left it in a working state with many good features, he came and did it again.
In the past, he added physical videomode setting, which has no sense in current displays, only to improve his own usercase, and opened a can of worms because he didn't test software beyond his own. I had to work to do many fixes after that incomplete implementation, and the KMSDRM backend became way more complicated because of videomode management, when it was a very simple backend before.

Now, after a while without looking at SDL2 commits, I update my builds and find NOTHING works. Just imagine my feelings.
I will moderate my tone, really. But please try all the software I have mentioned: it's not working right at all after his commits, which again are only tailored for a particular piece of closed-source software.

**Comment by slime73** [comment]
Are there certain steps that can be run (for example one of the SDL test programs with specific parameters) to verify that at least the majority of scenarios involving fullscreen work as expected in a certain backend?

This isn't a question directed at a specific person necessarily, it just seems like now's a good time to put together a list of things that people can quickly run FS-related tests on - and then check expected results against what they see - without needing familiarity with third party apps.

**Comment by vanfanel** [comment]
@slime73 there are not "complete" tests as far as I know. The current tests start in windowed mode, but none of them start in windowed mode and then change to fullscreen, or viceversa.
Also, there's no "readability" tests in the sense that a 320x240 window on a 4K display is UNREADABLE unless every window is forced to fill the screen (it's still s windowed window, but fills the screen: not all windows are forced to be fullscreen), as it was done before. These are usability concepts that no one has the right to alter on his own private needs.

**Comment by slime73** [comment]
> The current tests start in windowed mode, but none of them start in windowed mode and then change to fullscreen, or viceversa, for example.

The tests don't do that automatically but I glanced at the test code and it looks like it has a way to do that via hotkeys. I suppose that just underscores my theory that good documentation for what to test and how to test it could be a big help. :)

**Comment by icculus** [comment]
> The tests don't do that automatically but I glanced at the test code and it looks like it has a way to do that via hotkeys. I suppose that just underscores my theory that good documentation for what to test and how to test it could be a big help. :)

I'm going to open a separate issue about this; a small README about porting to new targets (or implementing new backends), and how to use the test apps to cover basic functionality, would be useful.

**Comment by icculus** [comment]
> Now, after a while without looking at SDL2 commits, I updated my SDL2 builds and found NOTHING works again. Just imagine my feelings.

I totally understand, because this happens to me sometimes too, in this and other projects.

But all active open source projects are fluid and changes are made by those participating, and universally when I've had those feelings, it was because I wandered away from that code and others took over.

I have not yet really dove into these patches, so I can't comment on the quality of them, but @inolen is correct: there is a mode setting API in SDL, and KMSDRM lets you set those modes, so there is no reason to not implement it so that KMSDRM exposes the functionality that SDL promises.

Added complexity in the code is unfortunate but it is also a way of life in SDL's internals. These things _are_ complex.

Let's figure out what new problems are happening and fix them, but being able to set modes isn't an unreasonable feature to offer.

**Comment by inolen** [comment]
In case anyone else is digging into the code here, I've pushed two more changes to https://github.com/inolen/SDL/commits/kmsdrm_regressions.

The first now always avoid overriding the mode requested with KMSDRM_SetDisplayMode when recreating surfaces to avoid surprises when toggling fullscreen.
The second fixes a crash I ran into due to the SetWindowSize / SetWindowFullscreen paths not deferring surface recreation.

These are unrelated to the issues being reported with PoP / sm64ex, so I'll hold off on PR'ing until that's sorted.

**Comment by Cacodemon345** [comment]
> Also, looking at the #4417 report, non-fullscreen windows seem also broken: They must fill the whole screen as they did before, because reading text on a 320x240 window on a 4K display is almost impossible and NOT convenient at all.

Can it be made it so that it will choose the next higher video mode available after the window size without scaling to the whole screen as a decent compromise? It would ensure programs can resize their window without screens getting cut off and would also solve the issue the PR author was trying to solve.

**Comment by vanfanel** [comment]
@Cacodemon345 There's no scaling done: what's done is that a matching (if possible) video mode is chosen. Why would the immediate higher mode be a better option than a matching mode?

**Comment by Cacodemon345** [comment]
Actually, what I meant is to center the SDL2 window when it is created instead of creating at top-left of the screen.

**Comment by LilaQ** [comment]
Great show, really enjoyed it. The clown was annoying though.

**Comment by Cacodemon345** [comment]
I already saw enough drama elsewhere outside SDL2 already. Let's stop adding fuel to the fire.

**Comment by icculus** [comment]
> Great show, really enjoyed it. The clown was annoying though.

This is a good way to get blocked at the project level. Please don't wander into a conversation and lob an insult at a contributor. Not only is it unproductive, it's deeply immature.

I shouldn't have to explain this. And next time I won't.
